internal/controller/clusterinstance/validator.go

```diff
@@ -100,21 +59,49 @@ func Validate(ctx context.Context, c client.Client, clusterInstance *v1alpha1.Cl
 				Name:      extraManifestRef.Name,
 				Namespace: clusterInstance.Namespace,
 			}, cm); err != nil {
-				return fmt.Errorf("failed to retrieve ExtraManifest: %s in namespace %s, err: %w", extraManifestRef.Name, clusterInstance.Namespace, err)
+				return fmt.Errorf(
+					"failed to retrieve ExtraManifest: %s in namespace %s, err: %w",
+					extraManifestRef.Name, clusterInstance.Namespace, err)
 			}
 		}
 	}
 
-	numControlPlaneAgents := 0
-	numWorkerAgents := 0
+	// Check that node BMC secrets exist in namespace
 	for _, node := range clusterInstance.Spec.Nodes {
+		bmcSecret := &corev1.Secret{}
+		if err := c.Get(ctx, types.NamespacedName{
+			Name:      node.BmcCredentialsName.Name,
+			Namespace: clusterInstance.Spec.ClusterName},
+			bmcSecret); err != nil {
+			return fmt.Errorf(
+				"failed to validate BMC credentials: %s in namespace %s [Node: Hostname=%s], err: %w",
+				node.BmcCredentialsName.Name, clusterInstance.Spec.ClusterName, node.HostName, err)
+		}
+	}
 
-		if node.Role == "master" {
-			numControlPlaneAgents++
-		} else if node.Role == "worker" {
-			numWorkerAgents++
+	// validation succeeded
+	return nil
+}
+
+func validateTemplateRefs(ctx context.Context, c client.Client, clusterInstance *v1alpha1.ClusterInstance) error {
+
+	// Check the cluster-level template references are defined
+	if (clusterInstance.Spec.TemplateRefs == nil) || (len(clusterInstance.Spec.TemplateRefs) < 1) {
+		return fmt.Errorf("missing cluster-level TemplateRefs")
+	}
+	// Verify that the cluster-level TemplateRefs exist
+	for _, templateRef := range clusterInstance.Spec.TemplateRefs {
+		cm := &corev1.ConfigMap{}
+		if err := c.Get(ctx, types.NamespacedName{
+			Name:      templateRef.Name,
+			Namespace: templateRef.Namespace,
+		}, cm); err != nil {
```

Comment:
I also don't live the formatting here.
Maybe split the `NamespacedName` struct into a variable and one-line the conditional.

Reply:
Nice suggestion, changed them all!